### PR TITLE
refactor: ASF only dependencies on base.txt

### PR DESCRIPTION
**Pull request by dpgaspar**
### SUMMARY
Removes all non base dependencies from `requirements/base.[in,txt]`. `base` is generated strictly from `setup.py`, where all dependencies are considered safe from a licensing perspective.

Basically added `development.in` has another dependency layer

Blank diagram

base.in: base pinned dependencies based on `setup.py` Apache license compliant only
development.in: Adds extra necessary dependencies for development
integration.in: Add dependencies for CI
docker.in: Docker extra requirements
local.in: hook to add local development custom extras
testing.in: Add dependencies for testing (CI and local)

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: requirements/docker.txt [hunk 1]
```diff
@@ -9,6 +9,7 @@
 -e file:.                 # via -r requirements/base.in
 gevent==20.6.2            # via -r requirements/docker.in
 greenlet==0.4.16          # via gevent
+psycopg2-binary==2.8.6    # via -r requirements/docker.in
```
Comment (robdiciuccio) [inline requirements/docker.txt:12]: My understanding is that the Docker build can't contain `psycopg2` due to licensing.

**Comment by robdiciuccio** [comment]
PR for fixing FOSSA license checks: https://github.com/apache/incubator-superset/pull/10848

File: requirements/docker.txt (hunk 1 above)
Comment (craig-rueda) [reply requirements/docker.txt:12]: This is only used to build Docker, which isn't actually pushed anywhere.

File: requirements/development.in [hunk 2]
```diff
@@ -14,6 +14,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+-r base.txt
```
Comment (villebro) [inline requirements/development.in:17]: Why is this needed? As development dependencies might be incompatible with the pinnings in `base.txt`, this might cause a clash.

File: requirements/development.in [hunk 3]
```diff
@@ -14,6 +14,8 @@
...
+-e file:.[cors,druid,hive,mysql,postgres,thumbnails]
```
Comment (villebro) [inline requirements/development.in:18]: Should we add presto here?

**Review by villebro — COMMENTED** [review]
A few comments

File: requirements/development.in (hunk 3 above)
Comment (dpgaspar) [reply requirements/development.in:17]: debatable, because we are already using local dev docker based on development.txt and it already has lot's of "unnecessary" dependencies there

**Review by john-bodley — COMMENTED** [review]
Sorry there’s a few things I don’t understand:

1. Why is `mysqlclient` not allowed in `base.txt` but is allowed in `setup.py`? Both files are included in the manifest as far as I’m aware.

2. I’m not sure why `integration.in` was incorporated into `development.in`. By definition this should contain the subset of packages required only to define the base integration environment for CI et al. and should be agnostic of Superset.

**Comment by dpgaspar** [comment]
> Sorry there’s a few things I don’t understand:
> 
> 1. Why is `mysqlclient` not allowed in `base.txt` but is allowed in `setup.py`? Both files are included in the manifest as far as I’m aware.
> 2. I’m not sure why `integration.in` was incorporated into `development.in`. By definition this should contain the subset of packages required only to define the base integration environment for CI et al. and should be agnostic of Superset.

1 - `mysqlclient` is allowed on `setup.py` has an extra requirement, so our goal here is to make a Apache license compliant requirements file 

2 - Yes that's a good point, I can re include `integration.in` has it was, and still have a `development.in` with all the necessary extras. 

Extra question: Should we base `development.in` on `base.in` or branch it directly from `setup.py` to include the extras (like it's currently done on this PR) ?

**Comment by john-bodley** [comment]
Ideally it should be based on `base.in` however I believe having multiple `-e file:.` in the dependency chain is problematic.

File: requirements/development.in [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+-r base.in
+-e file:.[cors,druid,hive,mysql,postgres,thumbnails]
+flask-cors>=2.0.0
```
Comment (dpgaspar) [inline requirements/development.in:20]: This is a bit of an hack to force `pip-compile-multi` to generate the necessary extra dependency locks. Since `-e file:.[cors,druid,hive,mysql,postgres,thumbnails]` was not being propagated to `development.txt`

Comment (john-bodley) [reply requirements/development.in:20]: All I see in `development.txt` is `-e file:.`. The fact that you're explicitly enumerating these requirements in `development.in` should be more than suffice to generate the desired outcome in `development.txt`.

Comment (dpgaspar) [reply requirements/development.in:19]: right, updated

**Review by john-bodley — APPROVED** [review]

**Review by villebro — APPROVED** [review]
LGTM, let's hope Python gets dependencies right sometime soon!
